tests/tests/integration_tests.rs

```diff
@@ -256,3 +256,26 @@ async fn server_should_be_able_to_close_subscriptions() {
 
 	assert!(matches!(res, Err(Error::SubscriptionClosed(_))));
 }
+
+#[tokio::test]
+async fn ws_close_pending_subscription_when_server_terminated() {
+	let (server_addr, mut handle) = websocket_server_with_subscription().await;
+	let server_url = format!("ws://{}", server_addr);
+
+	let c1 = WsClientBuilder::default().build(&server_url).await.unwrap();
+
+	let mut sub: Subscription<String> =
+		c1.subscribe("subscribe_hello", JsonRpcParams::NoParams, "unsubscribe_hello").await.unwrap();
+
+	assert!(matches!(sub.next().await, Ok(Some(_))));
+
+	handle.stop().await.unwrap();
+	handle.wait_for_stop().await;
```

Comment:
Offhand, it looks like wait_for_stop won't wait for existing connections to close as well as it'll just wait for the loop accepting new connections to end. Is this OK?

Reply:
right good catch, I think the `stop mutex` must be passed the background thread too for this to work properly and only be unlocked until all connections are closed.

Reply:
Instead of a mutex, I think you'd need a RwLock which acquired `read`'s on the various background tasks and writes on `wait_for_stop`.

Alternately you could do what RwLock does internally using Semaphores (create a new Semaphore with say u32::MAX permits, and then acquire a permit for each background task, and on `wait_for_stop`, acquire_many u32::MAX permits (which will block until all read permits are handed back when the reads finish)